core/repository/sparql/src/main/java/org/eclipse/rdf4j/repository/sparql/SPARQLConnection.java

```diff
@@ -827,39 +819,71 @@ public boolean isActive() throws UnknownTransactionStateException, RepositoryExc
 		}
 	}
 
+	@Override
+	protected void addWithoutCommit(Statement st, Resource... contexts)
+			throws RepositoryException {
+		flushPendingRemoves();
+		if (contexts.length == 0) {
+			pendingAdds.add(st);
+		} else {
+			pendingAdds.add(st.getSubject(), st.getPredicate(), st.getObject(), contexts);
+		}
+	}
+
 	@Override
 	protected void addWithoutCommit(Resource subject, IRI predicate, Value object, Resource... contexts)
 			throws RepositoryException {
-		ValueFactory f = getValueFactory();
+		flushPendingRemoves();
+		pendingAdds.add(subject, predicate, object, contexts);
+	}
 
-		Statement st = f.createStatement(subject, predicate, object);
+	private void flushPendingRemoves() {
+		if (!pendingRemoves.isEmpty()) {
+			for (Resource context : pendingRemoves.contexts()) {
+				String sparqlCommand = createDeleteDataCommand(pendingRemoves.getStatements(null, null, null, context),
+						context);
+				sparqlTransaction.append(sparqlCommand);
+				sparqlTransaction.append("; ");
+			}
+			pendingRemoves.clear();
+		}
+	}
 
-		List<Statement> list = new ArrayList<>(1);
-		list.add(st);
-		String sparqlCommand = createInsertDataCommand(list, contexts);
+	private void flushPendingAdds() {
+		if (!pendingAdds.isEmpty()) {
+			for (Resource context : pendingAdds.contexts()) {
+				String sparqlCommand = createInsertDataCommand(pendingAdds.getStatements(null, null, null, context),
+						context);
+				sparqlTransaction.append(sparqlCommand);
+				sparqlTransaction.append("; ");
+			}
+			pendingAdds.clear();
```

Comment:
I think .contexts() upgrades the DynamicModel to a HashModel. This is probably not going to be what is performance critical in this situation though. I'm also unsure if .clear() downgrades the model again.

Reply:
> I'm also unsure if .clear() downgrades the model again.

That might be an interesting improvement to DynamicModel. I mostly just used a model here because it's a simple way to keep track of statements per context. I could perhaps also use a Multimap of some sort if this turns out to be heavy.

Reply:
For now I've changed it slightly to just drop the model and create a new empty one when flushing.